integration/upload_test.go

```diff
@@ -0,0 +1,257 @@
+		require.NoErrorf(suite.T(), err, msgErrorExecutingOperation, operationStart)
+	}
+
+	statuses := []interface{}{}
+	err = util.ProcessWSMessages(suite.Cfg, connEvents,
+		func(msg *protocol.Envelope) (bool, error) {
+			if msg.Topic.String() == topicModified && msg.Path == pathLastUpload {
+				statuses = append(statuses, msg.Value)
+				return isTerminal(msg.Value), nil
+			}
+			return false, fmt.Errorf(msgUnexpectedValue, msg.Value)
+		})
+	require.NoError(suite.T(), err, "error processing upload status events")
+
+	lastUploadProgress := 0
+	for ind, status := range statuses {
+		uploadStatus := uploadStatus{}
+		err := parseEventValue(status, &uploadStatus)
+		require.NoErrorf(suite.T(), err, "cannot convert %v to upload request, will ignore the event", status)
+		suite.T().Logf("upload status event(%v)", uploadStatus)
+		require.True(suite.T(), uploadStatus.Progress >= lastUploadProgress && uploadStatus.Progress <= 100,
+			"upload status progress should be non-decreasing and in range [0, 100]")
+		lastUploadProgress = uploadStatus.Progress
+		if ind < len(statuses)-1 {
+			require.Equal(suite.T(), client.StateUploading, uploadStatus.State, "wrong transitional upload state")
+		} else {
+			require.Equal(suite.T(), client.StateSuccess, uploadStatus.State, "wrong final upload state")
+		}
+	}
+	for _, filePath := range files {
+		startID, ok := fileIDs[filePath]
+		require.True(suite.T(), ok, fmt.Sprintf("no upload request event for %s", filePath))
+		content, err := testUpload.download(startID)
+		require.NoError(suite.T(), err, fmt.Sprintf("file %s not uploaded", filePath))
+		suite.compareContent(filePath, content)
+	}
+}
+
+func (suite *fileUploadSuite) testUpload(testUpload upload) {
+	files, err := createTestFiles(suite.uploadCfg.UploadDir)
+	defer suite.deleteTestFiles(files)
+	require.NoError(suite.T(), err, "creating test files failed")
+
+	requestedFiles := suite.triggerUploads()
+	defer testUpload.removeUploads()
+
+	suite.startUploads(testUpload, requestedFiles, files)
+}
+
+func (suite *fileUploadSuite) checkUploadDir() {
+	files, err := os.ReadDir(suite.uploadCfg.UploadDir)
+	if err != nil {
+		suite.T().Fatalf("upload dir %s cannot be read - %v", suite.uploadCfg.UploadDir, err)
+	}
+	if len(files) > 0 {
+		suite.T().Fatalf("upload dir %s must be empty", suite.uploadCfg.UploadDir)
+	}
+}
+
+func (suite *fileUploadSuite) compareContent(filePath string, received []byte) {
+	expected, err := ioutil.ReadFile(filePath)
+	require.NoError(suite.T(), err, fmt.Sprintf("cannot read file %s", filePath))
+	require.Equal(suite.T(), string(expected), string(received), fmt.Sprintf("uploaded content of file %s differs from original", filePath))
+}
+
+func (suite *fileUploadSuite) deleteTestFiles(files []string) {
+	for _, file := range files {
+		if err := os.Remove(file); err != nil {
+			suite.T().Logf("error deleting test file %s, please consider removing it manually(%v)", file, err)
```

Comment:
`please consider removing it manually` - it's an integration test that will be automated. I would suggest not to use such hints.